[FEAT] Updates State Management Docs (#1067)
ember-learn/guides-source: guides/release/state-management/computed-properties.md

Computed properties are a concept in Ember prior to Octane. They serve the same
purpose as tracked properties and native getters, allowing users to respond to

> changes, derive state, and ultimately update the DOM. They also have built-in

Reviewer: Pulling in a prior definition from [here](https://guides.emberjs.com/release/object-model/computed-properties/#toc_computed-properties-only-recompute-when-they-are-consumed)

```suggestion
changes, derive state, and ultimately update the DOM. 
In a nutshell, computed properties let you declare functions as properties.
You create one by defining a computed property as a function, which Ember will automatically call when you ask for the property. 
They also have built-in
```

caching to prevent having to perform expensive calculations more than once.

> While computed properties are no longer the recommended default, it's likely

Reviewer: ```suggestion
While computed properties are no longer the recommended default, it is likely
```

that you may encounter them in code that hasn't been updated to tracked
properties just yet, either in existing applications or in the wider Ember
ecosystem, so this guide exists both to describe how they work and can be used,

> and how they work with tracked properties.

Reviewer: What do you think about explicitly providing guidance to beginners here?

```suggestion
and how they work with tracked properties.
If you are just learning Ember for the first time, you can skip this section.
```

decorate standard computed property getters and setters:

```javascript
import { computed, set } from '@ember/object';

class Person {
  constructor(firstName, lastName) {
    set(this, 'firstName', firstName);
    set(this, 'lastName', lastName);
  }

  @computed('firstName', 'lastName')
  get fullName() {
    return `${this.firstName} ${this.lastName}`;
  }
}

let ironMan = new Person('Tony', 'Stark');

ironMan.fullName; // "Tony Stark"
```

This computed property works just like a normal getter/setter, with two key
differences:

1. It will cache its value by default, and it will only update that value if its
   _dependencies_, in this case the `firstName` and `lastName` properties,
   change.

   ```javascript
   class Counter {
     _count = 0;

     @computed
     get count() {
       console.log('counted!');
       return this._count;
     }
   }

   let counter = new Counter();

   counter.count; // logs 'counted!'
   counter.count; // logs nothing, the values was cached and hasn't updated
   ```

2. It will notify other "watchers", such as other computed properties and
   templates, if any of its dependencies has updated and it needs to be
   recalculated.

> 

Reviewer: I think maybe we need to add a third point. One of the most significant gotchas of CPs for new devs is that they only run if you are using them in the template. We should state this explicitly somewhere. We can borrow from [explanations in the existing guides](https://guides.emberjs.com/release/object-model/computed-properties/#toc_computed-properties-only-recompute-when-they-are-consumed):

```suggestion
3. A computed property will only recompute its value when it is consumed. Properties are consumed in two ways: by being accessed, like `ironMan.fullName`, or by being referenced in a handlebars template that is currently being rendered, like `{{this.ironMan.fullName}}`.

```

  special truncated syntax as shorthand:

  ```js
  class Profile {
    constructor(user) {
      set(this, 'user', user);
    }

    @computed('user.{firstName,lastName}')
    get userName() {
      return `${this.user.firstName} ${this.user.lastName}`;
    }
  }
  ```

  Note that no spaces are allowed in this truncated syntax, Ember will assert if
  you place any inside of it.

- **Array dependencies.** It's possible to depend on an array, and the items in
  the array, by watching the `[]` property on the array:

  ```js
  class Person {
    constructor(friends = []) {
      set(this, 'friends', friends);
    }

    @computed('friends.[]')
    get friendNames() {
      return this.friends.map(friend => friend.name);
    }
  }
  ```

  You can also depend directly on a _property_ of each item in the array using
  `@each` syntax:

  ```js
  class Person {
    constructor(friends = []) {
      set(this, 'friends', friends);
    }

    @computed('friends.@each.name')
    get friendNames() {
      return this.friends.map(friend => friend.name);
    }
  }
  ```

  However, you cannot _chain_ on these properties, as it is a performance

>   pitfall. You can only do 1 level of `@each` watching.

Reviewer: ```suggestion
  pitfall. You can only do one level of `@each` watching.
```

### Defining Setters

> If you define a setter for your computed property, it'll work just like a normal

Reviewer: ```suggestion
If you define a setter for your computed property, it will work just like a normal
```

Feedback from non-native-English-speakers is that contractions can be difficult. We do use `let's` liberally and I don't usually remove it.
